### PR TITLE
Add template for Tensor Parallel (DeepSpeed AutoTP) + Ray Train

### PR DESCRIPTION
**Pull request by tohtana**
This PR adds a template that demonstrates how to use Ray Train with DeepSpeed AutoTP for tensor parallelism.

At a high level, this template covers:

- A hands-on example of fine-tuning a language model using 2D parallelism (Tensor Parallelism + Data Parallelism).
- Setting up DeepSpeed AutoTP for automatic model sharding across GPUs.
- TP-aware data loading to ensure correct gradient computation across tensor-parallel ranks.
- Saving and loading distributed model checkpoints with Ray Train.
- Key configurations for scaling to larger models (adjusting TP/DP sizes, ZeRO stages).

Tested environment: 1 CPU head node + 1 worker node with 4 NVIDIA T4 GPUs.

### CODE REVIEW

File: doc/source/train/examples/pytorch/tensor_parallel_autotp/README.ipynb [hunk 1]
```diff
@@ -0,0 +1,857 @@
...
+        "\n",
+        "**Tensor Parallelism (TP)** shards model weights across multiple GPUs, enabling training of models that are too large to fit on a single GPU. Combined with **Data Parallelism (DP)**, this creates a powerful **2D parallelism** strategy that scales efficiently to many GPUs.\n",
+        "\n",
+        "This tutorial provides a step-by-step guide covering:\n",
+        "\n",
+        "- Understanding 2D parallelism (Tensor Parallelism + Data Parallelism)\n",
+        "- Setting up DeepSpeed AutoTP for automatic model sharding\n",
+        "- Combining with ZeRO optimization for data parallelism\n",
+        "- TP-aware data loading to ensure correct gradient computation\n",
+        "- Distributed checkpointing with Ray Train\n",
```
Comment (liulehui) [inline doc/source/train/examples/pytorch/tensor_parallel_autotp/README.ipynb]: ditto with the other one to align order with the steps below.

1. understanding 2D parallelism.
2. Configure DataLoader for TP-aware sharding
3. Set up Model with DeepSpeed AutoTP and ZeRO
4. checkpointing
5. Distributed Training with Ray Train.

File: doc/source/train/examples/pytorch/tensor_parallel_autotp/README.ipynb [hunk 2]
```diff
@@ -0,0 +1,857 @@
...
+        "\n",
+        "    # [6] Build DeepSpeed config\n",
+        "    batch_size = config.get(\"batch_size\", 1)\n",
+        "    effective_dp = dp_size if tp_size > 1 else world_size\n",
+        "\n",
+        "    ds_config = {\n",
+        "        \"train_batch_size\": batch_size * effective_dp,\n",
+        "        \"train_micro_batch_size_per_gpu\": batch_size,\n",
+        "        \"gradient_accumulation_steps\": 1,\n",
+        "        \"gradient_clipping\": config.get(\"max_grad_norm\", 1.0),\n",
```
Comment (liulehui) [inline doc/source/train/examples/pytorch/tensor_parallel_autotp/README.ipynb]: these can all from the `config` passed to trainer,
how about we do `gradient_accumulation_steps`: config.get("gradient_accumulation_steps", 1) to allow user to pass in in the `train_loop_config` ?

**Review by liulehui — COMMENTED** [review]
❤️

File: doc/source/train/examples/pytorch/tensor_parallel_autotp/README.ipynb [hunk 3]
```diff
@@ -0,0 +1,857 @@
...
+        "    return DataLoader(tokenized, batch_size=batch_size, sampler=sampler, drop_last=True)"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {},
+      "source": [
+        "## 3. Model Parallelization with DeepSpeed AutoTP\n",
+        "\n",
+        "DeepSpeed AutoTP provides automatic tensor parallelism through:\n",
```
Comment (JasonLi1909) [inline doc/source/train/examples/pytorch/tensor_parallel_autotp/README.ipynb]: Link to some api docs could be helpful here

**Review by JasonLi1909 — COMMENTED** [review]
Great template! Is there a case you might want to use Pytorch DTensor + FSDP over Deepspeed AutoTP + Zero3 for tp, dp or is AutoTP always more convenient? Could be good to have a brief explanation covering this.

File: doc/source/train/examples/pytorch/tensor_parallel_autotp/README.ipynb (hunk 3 above)
Comment (tohtana) [reply doc/source/train/examples/pytorch/tensor_parallel_autotp/README.ipynb]: Added a link to [AutoTP doc](https://deepspeed.readthedocs.io/en/latest/training.html#automatic-tensor-parallel-training).

**Comment by tohtana** [comment]
> Is there a case you might want to use Pytorch DTensor + FSDP over Deepspeed AutoTP + Zero3 for tp, dp or is AutoTP always more convenient? Could be good to have a brief explanation covering this.

Thank you for your feedback! Added a paragraph to explain pros/cons of these two approaches.

File: doc/source/train/examples/pytorch/tensor_parallel_autotp/README.ipynb (hunk 2 above)
Comment (tohtana) [reply doc/source/train/examples/pytorch/tensor_parallel_autotp/README.ipynb]: Fixed as suggested (also add zero_stage in the config).

File: doc/source/train/examples/pytorch/tensor_parallel_autotp/README.ipynb (hunk 1 above)
Comment (tohtana) [reply doc/source/train/examples/pytorch/tensor_parallel_autotp/README.ipynb]: Thank you, reordered the items.

**Review by matthewdeng — APPROVED** [review]
